File: r/pragmastat/tests/testthat/test-assumptions.R
# Assumption violation conformance tests
#
# These tests verify that assumption violations are reported correctly and
# consistently across all languages. The test data is loaded from shared
# JSON files in tests/assumptions/.

# Helper function to parse special values from JSON
parse_value <- function(v) {
  if (is.numeric(v)) {
    return(v)
  }
  if (is.character(v)) {
    switch(v,
      "NaN" = NaN,
      "Infinity" = Inf,
      "-Infinity" = -Inf,
      stop(paste("Unknown string value:", v))
    )
  } else {
    stop(paste("Unexpected value type:", typeof(v)))
  }
}

# Helper function to parse an array of values from nested list structure
parse_array <- function(arr) {
  if (is.null(arr)) {
    return(numeric(0))
  }
  if (length(arr) == 0) {
    return(numeric(0))
  }
  # Handle list of values (may contain special strings like "NaN")
  sapply(arr, parse_value)
}

# Function dispatch: maps function names to actual implementations
call_function <- function(func_name, x, y) {
  switch(func_name,
    "Center" = center(x),
    "Ratio" = ratio(x, y),
    "RelSpread" = rel_spread(x),
    "Spread" = spread(x),
    "Shift" = shift(x, y),
    "AvgSpread" = avg_spread(x, y),
    "Disparity" = disparity(x, y),
    stop(paste("Unknown function:", func_name))
  )
}

test_that("assumption violations are correct", {
  # Find repository root by looking for testthat.R
  current_dir <- getwd()
  repo_root <- current_dir
  while (!file.exists(file.path(repo_root, "testthat.R"))) {
    parent_dir <- dirname(repo_root)
    if (parent_dir == repo_root) {
      stop(paste0("Could not find repository root (testthat.R not found); current dir is ", getwd()))
    }
    repo_root <- parent_dir
  }

  assumptions_dir <- file.path(repo_root, "tests", "assumptions")

  # Load manifest
  manifest <- jsonlite::fromJSON(file.path(assumptions_dir, "manifest.json"))

  total_tests <- 0
  passed_tests <- 0

  # Run each suite
  for (i in seq_len(nrow(manifest$suites))) {
    suite_entry <- manifest$suites[i, ]
    suite_path <- file.path(assumptions_dir, suite_entry$file)
    # Use simplifyVector = FALSE to preserve structure for special values
    suite <- jsonlite::fromJSON(suite_path, simplifyVector = FALSE)

    for (j in seq_along(suite$cases)) {
      test_case <- suite$cases[[j]]
      test_name <- paste0(suite$suite, "/", test_case$name)
      total_tests <- total_tests + 1

      x <- parse_array(test_case$inputs$x)
      y <- parse_array(test_case$inputs$y)

      expected_id <- test_case$expected_violation$id
      expected_subject <- test_case$expected_violation$subject

      # Get function name
      func_name <- test_case[["function"]]

      # Try to call the function and catch assumption violations
      result <- tryCatch({
        call_function(func_name, x, y)
        list(success = TRUE, error = NULL)
      }, assumption_error = function(e) {
        list(success = FALSE, error = e)
      }, error = function(e) {
        # Other errors
        list(success = FALSE, error = e)
      })

      if (result$success) {
        fail(paste0(test_name, ": Expected violation ", expected_id, "(", expected_subject, ") but got success"))
        next
      }

      err <- result$error
      if (!is_assumption_error(err)) {
        fail(paste0(test_name, ": Expected AssumptionError but got ", class(err)[1], ": ", conditionMessage(err)))
        next
      }

      actual_id <- err$violation$id
      actual_subject <- err$violation$subject

      expect_equal(actual_id, expected_id,
        info = paste0(test_name, ": Expected id=", expected_id, ", got ", actual_id))
      expect_equal(actual_subject, expected_subject,
        info = paste0(test_name, ": Expected subject=", expected_subject, ", got ", actual_subject))

      passed_tests <- passed_tests + 1
    }
  }

  message(paste0("\nAssumption Tests: ", passed_tests, "/", total_tests, " passed"))
})

File: py/tests/test_assumptions.py
"""Assumption violation conformance tests.

These tests verify that assumption violations are reported correctly and
consistently across all languages. The test data is loaded from shared
JSON files in tests/assumptions/.
"""

import json
from pathlib import Path

import numpy as np
import pytest

from pragmastat import (
    avg_spread,
    center,
    center_bounds,
    disparity,
    median_bounds,
    ratio,
    rel_spread,
    shift,
    signed_rank_margin,
    spread,
)
from pragmastat.assumptions import AssumptionError


def find_repo_root() -> Path:
    """Finds the repository root by looking for CITATION.cff."""
    current = Path(__file__).resolve()
    while current != current.parent:
        if (current / "CITATION.cff").exists():
            return current
        current = current.parent
    raise RuntimeError("Could not find repository root (CITATION.cff not found)")


def parse_value(v) -> float:
    """Parses a JSON value into a float, handling special values."""
    if isinstance(v, (int, float)):
        return float(v)
    if isinstance(v, str):
        if v == "NaN":
            return float("nan")
        if v == "Infinity":
            return float("inf")
        if v == "-Infinity":
            return float("-inf")
        raise ValueError(f"Unknown string value: {v}")
    raise TypeError(f"Unexpected value type: {type(v)}")


def parse_array(arr: list | None) -> np.ndarray:
    """Parses a JSON array into a numpy array."""
    if arr is None:
        return np.array([])
    return np.array([parse_value(v) for v in arr])


def call_function(func_name: str, inputs: dict) -> None:
    """Dispatches to the appropriate estimator function."""
    x = parse_array(inputs.get("x"))
    y = parse_array(inputs.get("y"))
    misrate_raw = inputs.get("misrate")
    misrate = parse_value(misrate_raw) if misrate_raw is not None else None
    n = inputs.get("n")
    seed = inputs.get("seed")

    dispatch = {
        "Center": lambda: center(x),
        "Ratio": lambda: ratio(x, y),
        "RelSpread": lambda: rel_spread(x),
        "Spread": lambda: spread(x),
        "Shift": lambda: shift(x, y),
        "AvgSpread": lambda: avg_spread(x, y),
        "Disparity": lambda: disparity(x, y),
        "MedianBounds": lambda: median_bounds(x, misrate),
        "CenterBounds": lambda: center_bounds(x, misrate),
        "SignedRankMargin": lambda: signed_rank_margin(n, misrate),
    }
    if func_name not in dispatch:
        raise ValueError(f"Unknown function: {func_name}")
    dispatch[func_name]()


def load_assumption_test_cases():
    """Loads all assumption test cases from the shared test data."""
    repo_root = find_repo_root()
    assumptions_dir = repo_root / "tests" / "assumptions"

    # Load manifest
    manifest_path = assumptions_dir / "manifest.json"
    with open(manifest_path) as f:
        manifest = json.load(f)

    test_cases = []
    for suite_entry in manifest["suites"]:
        suite_path = assumptions_dir / suite_entry["file"]
        with open(suite_path) as f:
            suite = json.load(f)

        for case in suite["cases"]:
            test_cases.append(
                pytest.param(
                    case["function"],
                    case["inputs"],
                    case["expected_violation"]["id"],
                    id=f"{suite['suite']}/{case['name']}",
                )
            )

    return test_cases


@pytest.mark.parametrize(
    "func_name,inputs,expected_id",
    load_assumption_test_cases(),
)
def test_assumption_violation(func_name, inputs, expected_id):
    """Tests that the correct assumption violation is raised."""
    with pytest.raises(AssumptionError) as exc_info:
        call_function(func_name, inputs)

    err = exc_info.value
    assert err.violation.id.value == expected_id, (
        f"Expected id={expected_id}, got {err.violation.id.value}"
    )
